File: static/pages/news/examples/api-client-examples.py
"""
نمونه کدهای استفاده از API اخبار کریپتو
Crypto News API Client Examples in Python

این فایل شامل مثال‌های مختلف برای استفاده از API اخبار است
This file contains various examples for using the News API
"""

import requests
import json
from typing import List, Dict, Optional
from datetime import datetime


class CryptoNewsClient:
    """
    کلاس کلاینت برای دسترسی به API اخبار
    Client class for accessing the News API
    """
    
    def __init__(self, base_url: str = "http://localhost:3000"):
        """
        مقداردهی اولیه کلاینت
        Initialize the client
        
        Args:
            base_url: آدرس پایه سرور / Base URL of the server
        """
        self.base_url = base_url
        self.session = requests.Session()
        self.session.headers.update({
            'Accept': 'application/json',
            'User-Agent': 'CryptoNewsClient/1.0'
        })
    
    def get_all_news(self, limit: int = 100) -> List[Dict]:
        """
        دریافت تمام اخبار
        Get all news articles
        
        Example:
            >>> client = CryptoNewsClient()
            >>> articles = client.get_all_news(limit=50)
            >>> print(f"Found {len(articles)} articles")
        """
        url = f"{self.base_url}/api/news"
        params = {'limit': limit}
        
        try:
            response = self.session.get(url, params=params, timeout=10)
            response.raise_for_status()
            data = response.json()
            return data.get('articles', [])
        except requests.exceptions.RequestException as e:
            print(f"خطا در دریافت اخبار / Error fetching news: {e}")
            return []
    
    def get_news_by_sentiment(self, sentiment: str, limit: int = 50) -> List[Dict]:
        """
        دریافت اخبار بر اساس احساسات
        Get news by sentiment
        
        Args:
            sentiment: 'positive', 'negative', or 'neutral'
            limit: تعداد نتایج / Number of results
        
        Example:
            >>> client = CryptoNewsClient()
            >>> positive_news = client.get_news_by_sentiment('positive')
            >>> for article in positive_news[:5]:
            ...     print(article['title'])
        """
        url = f"{self.base_url}/api/news"
        params = {
            'sentiment': sentiment,
            'limit': limit
        }
        
        try:
            response = self.session.get(url, params=params, timeout=10)
            response.raise_for_status()
            data = response.json()
            articles = data.get('articles', [])
            
            # فیلتر سمت کلاینت / Client-side filter
            return [a for a in articles if a.get('sentiment') == sentiment]
        except requests.exceptions.RequestException as e:
            print(f"Error: {e}")
            return []
    
    def get_news_by_source(self, source: str, limit: int = 50) -> List[Dict]:
        """
        دریافت اخبار از یک منبع خاص
        Get news from a specific source
        
        Example:
            >>> client = CryptoNewsClient()
            >>> coindesk_news = client.get_news_by_source('CoinDesk')
        """
        url = f"{self.base_url}/api/news"
        params = {
            'source': source,
            'limit': limit
        }
        
        try:
            response = self.session.get(url, params=params, timeout=10)
            response.raise_for_status()
            data = response.json()
            return data.get('articles', [])
        except requests.exceptions.RequestException as e:
            print(f"Error: {e}")
            return []
    
    def search_news(self, keyword: str, limit: int = 100) -> List[Dict]:
        """
        جستجوی اخبار بر اساس کلمه کلیدی
        Search news by keyword
        
        Example:
            >>> client = CryptoNewsClient()
            >>> bitcoin_news = client.search_news('bitcoin')
            >>> print(f"Found {len(bitcoin_news)} articles about Bitcoin")
        """
        articles = self.get_all_news(limit)
        keyword_lower = keyword.lower()
        
        return [
            article for article in articles
            if keyword_lower in article.get('title', '').lower() or
               keyword_lower in article.get('content', '').lower()
        ]
    
    def get_latest_news(self, count: int = 10) -> List[Dict]:
        """
        دریافت آخرین اخبار
        Get latest news
        
        Example:
            >>> client = CryptoNewsClient()
            >>> latest = client.get_latest_news(5)
            >>> for article in latest:
            ...     print(f"{article['title']} - {article['published_at']}")
        """
        articles = self.get_all_news(limit=100)
        
        # مرتب‌سازی بر اساس تاریخ انتشار / Sort by publish date
        sorted_articles = sorted(
            articles,
            key=lambda x: x.get('published_at', ''),
            reverse=True
        )
        
        return sorted_articles[:count]
    
    def get_news_statistics(self) -> Dict:
        """
        دریافت آمار اخبار
        Get news statistics
        
        Returns:
            Dictionary containing statistics
        
        Example:
            >>> client = CryptoNewsClient()
            >>> stats = client.get_news_statistics()
            >>> print(f"Total articles: {stats['total']}")
            >>> print(f"Positive: {stats['positive']}")
            >>> print(f"Negative: {stats['negative']}")
        """
        articles = self.get_all_news()
        
        stats = {
            'total': len(articles),
            'positive': sum(1 for a in articles if a.get('sentiment') == 'positive'),
            'negative': sum(1 for a in articles if a.get('sentiment') == 'negative'),
            'neutral': sum(1 for a in articles if a.get('sentiment') == 'neutral'),
            'sources': len(set(a.get('source', {}).get('title', '') for a in articles))
        }
        
        return stats


# ==============================================================================
# مثال‌های استفاده / Usage Examples
# ==============================================================================

def example_1_basic_usage():
    """مثال ۱: استفاده ساده / Example 1: Basic Usage"""
    print("=" * 60)
    print("مثال ۱: دریافت تمام اخبار / Example 1: Get All News")
    print("=" * 60)
    
    client = CryptoNewsClient()
    articles = client.get_all_news(limit=10)
    
    print(f"\nتعداد مقالات / Number of articles: {len(articles)}\n")
    
    for i, article in enumerate(articles[:5], 1):
        print(f"{i}. {article.get('title', 'No title')}")
        print(f"   منبع / Source: {article.get('source', {}).get('title', 'Unknown')}")
        print(f"   احساسات / Sentiment: {article.get('sentiment', 'neutral')}")
        print()


def example_2_sentiment_filtering():
    """مثال ۲: فیلتر بر اساس احساسات / Example 2: Sentiment Filtering"""
    print("=" * 60)
    print("مثال ۲: فیلتر اخبار مثبت / Example 2: Positive News Filter")
    print("=" * 60)
    
    client = CryptoNewsClient()
    positive_news = client.get_news_by_sentiment('positive', limit=50)
    
    print(f"\nاخبار مثبت / Positive news: {len(positive_news)}\n")
    
    for article in positive_news[:3]:
        print(f"✓ {article.get('title', 'No title')}")
        print(f"  {article.get('content', '')[:100]}...")
        print()


def example_3_keyword_search():
    """مثال ۳: جستجو با کلمه کلیدی / Example 3: Keyword Search"""
    print("=" * 60)
    print("مثال ۳: جستجوی بیت‌کوین / Example 3: Bitcoin Search")
    print("=" * 60)
    
    client = CryptoNewsClient()
    bitcoin_news = client.search_news('bitcoin')
    
    print(f"\nمقالات مرتبط با بیت‌کوین / Bitcoin articles: {len(bitcoin_news)}\n")
    
    for article in bitcoin_news[:5]:
        print(f"• {article.get('title', 'No title')}")


def example_4_statistics():
    """مثال ۴: آمار اخبار / Example 4: News Statistics"""
    print("=" * 60)
    print("مثال ۴: آمار اخبار / Example 4: Statistics")
    print("=" * 60)
    
    client = CryptoNewsClient()
    stats = client.get_news_statistics()
    
    print("\n📊 آمار / Statistics:")
    print(f"   مجموع مقالات / Total: {stats['total']}")
    print(f"   مثبت / Positive: {stats['positive']} ({stats['positive']/stats['total']*100:.1f}%)")
    print(f"   منفی / Negative: {stats['negative']} ({stats['negative']/stats['total']*100:.1f}%)")
    print(f"   خنثی / Neutral: {stats['neutral']} ({stats['neutral']/stats['total']*100:.1f}%)")
    print(f"   منابع / Sources: {stats['sources']}")


def example_5_latest_news():
    """مثال ۵: آخرین اخبار / Example 5: Latest News"""
    print("=" * 60)
    print("مثال ۵: آخرین اخبار / Example 5: Latest News")
    print("=" * 60)
    
    client = CryptoNewsClient()
    latest = client.get_latest_news(5)
    
    print("\n🕒 آخرین اخبار / Latest news:\n")
    
    for i, article in enumerate(latest, 1):
        published = article.get('published_at', '')
        if published:
            dt = datetime.fromisoformat(published.replace('Z', '+00:00'))
            time_str = dt.strftime('%Y-%m-%d %H:%M')
        else:
            time_str = 'Unknown time'
        
        print(f"{i}. {article.get('title', 'No title')}")
        print(f"   زمان / Time: {time_str}")
        print()


def example_6_advanced_filtering():
    """مثال ۶: فیلتر پیشرفته / Example 6: Advanced Filtering"""
    print("=" * 60)
    print("مثال ۶: فیلتر ترکیبی / Example 6: Combined Filters")
    print("=" * 60)
    
    client = CryptoNewsClient()
    
    # دریافت اخبار مثبت درباره اتریوم
    # Get positive news about Ethereum
    all_news = client.get_all_news(limit=100)
    
    filtered = [
        article for article in all_news
        if article.get('sentiment') == 'positive' and
           'ethereum' in article.get('title', '').lower()
    ]
    
    print(f"\nاخبار مثبت درباره اتریوم / Positive Ethereum news: {len(filtered)}\n")
    
    for article in filtered[:3]:
        print(f"✓ {article.get('title', 'No title')}")
        print(f"  منبع / Source: {article.get('source', {}).get('title', 'Unknown')}")
        print()


def main():
    """تابع اصلی / Main function"""
    print("\n" + "=" * 60)
    print("نمونه‌های استفاده از API اخبار کریپتو")
    print("Crypto News API Usage Examples")
    print("=" * 60 + "\n")
    
    try:
        # اجرای تمام مثال‌ها / Run all examples
        example_1_basic_usage()
        print("\n")
        
        example_2_sentiment_filtering()
        print("\n")
        
        example_3_keyword_search()
        print("\n")
        
        example_4_statistics()
        print("\n")
        
        example_5_latest_news()
        print("\n")
        
        example_6_advanced_filtering()
        
    except Exception as e:
        print(f"\nخطا / Error: {e}")
        print("لطفاً مطمئن شوید که سرور در حال اجرا است")
        print("Please make sure the server is running")


if __name__ == "__main__":
    main()

File: _hf_spaces/_ds_fixed_src/static/pages/news/examples/api-client-examples.js
/**
 * نمونه کدهای استفاده از API اخبار کریپتو
 * Crypto News API Client Examples in JavaScript/Node.js
 * 
 * این فایل شامل مثال‌های مختلف برای استفاده از API اخبار است
 * This file contains various examples for using the News API
 */

/**
 * کلاس کلاینت برای دسترسی به API اخبار
 * Client class for accessing the News API
 */
class CryptoNewsClient {
  /**
   * @param {string} baseUrl - آدرس پایه سرور / Base URL of the server
   */
  constructor(baseUrl = window.location.origin) {
    this.baseUrl = baseUrl;
  }

  /**
   * دریافت تمام اخبار
   * Get all news articles
   * 
   * @param {number} limit - تعداد نتایج / Number of results
   * @returns {Promise<Array>} آرایه مقالات / Array of articles
   * 
   * @example
   * const client = new CryptoNewsClient();
   * const articles = await client.getAllNews(50);
   * console.log(`Found ${articles.length} articles`);
   */
  async getAllNews(limit = 100) {
    try {
      const url = `${this.baseUrl}/api/news?limit=${limit}`;
      const response = await fetch(url);
      
      if (!response.ok) {
        throw new Error(`HTTP ${response.status}: ${response.statusText}`);
      }
      
      const data = await response.json();
      return data.articles || [];
    } catch (error) {
      console.error('خطا در دریافت اخبار / Error fetching news:', error);
      return [];
    }
  }

  /**
   * دریافت اخبار بر اساس احساسات
   * Get news by sentiment
   * 
   * @param {string} sentiment - 'positive', 'negative', or 'neutral'
   * @param {number} limit - تعداد نتایج / Number of results
   * @returns {Promise<Array>}
   * 
   * @example
   * const client = new CryptoNewsClient();
   * const positiveNews = await client.getNewsBySentiment('positive');
   * positiveNews.forEach(article => console.log(article.title));
   */
  async getNewsBySentiment(sentiment, limit = 50) {
    try {
      const url = `${this.baseUrl}/api/news?sentiment=${sentiment}&limit=${limit}`;
      const response = await fetch(url);
      
      if (!response.ok) {
        throw new Error(`HTTP ${response.status}`);
      }
      
      const data = await response.json();
      const articles = data.articles || [];
      
      // فیلتر سمت کلاینت / Client-side filter
      return articles.filter(a => a.sentiment === sentiment);
    } catch (error) {
      console.error('Error:', error);
      return [];
    }
  }

  /**
   * دریافت اخبار از یک منبع خاص
   * Get news from a specific source
   * 
   * @param {string} source - نام منبع / Source name
   * @param {number} limit - تعداد نتایج / Number of results
   * @returns {Promise<Array>}
   * 
   * @example
   * const client = new CryptoNewsClient();
   * const coinDeskNews = await client.getNewsBySource('CoinDesk');
   */
  async getNewsBySource(source, limit = 50) {
    try {
      const url = `${this.baseUrl}/api/news?source=${encodeURIComponent(source)}&limit=${limit}`;
      const response = await fetch(url);
      
      if (!response.ok) {
        throw new Error(`HTTP ${response.status}`);
      }
      
      const data = await response.json();
      return data.articles || [];
    } catch (error) {
      console.error('Error:', error);
      return [];
    }
  }

  /**
   * جستجوی اخبار بر اساس کلمه کلیدی
   * Search news by keyword
   * 
   * @param {string} keyword - کلمه کلیدی / Keyword
   * @param {number} limit - تعداد نتایج / Number of results
   * @returns {Promise<Array>}
   * 
   * @example
   * const client = new CryptoNewsClient();
   * const bitcoinNews = await client.searchNews('bitcoin');
   * console.log(`Found ${bitcoinNews.length} articles about Bitcoin`);
   */
  async searchNews(keyword, limit = 100) {
    const articles = await this.getAllNews(limit);
    const keywordLower = keyword.toLowerCase();
    
    return articles.filter(article => {
      const title = (article.title || '').toLowerCase();
      const content = (article.content || '').toLowerCase();
      return title.includes(keywordLower) || content.includes(keywordLower);
    });
  }

  /**
   * دریافت آخرین اخبار
   * Get latest news
   * 
   * @param {number} count - تعداد نتایج / Number of results
   * @returns {Promise<Array>}
   * 
   * @example
   * const client = new CryptoNewsClient();
   * const latest = await client.getLatestNews(5);
   * latest.forEach(article => {
   *   console.log(`${article.title} - ${article.published_at}`);
   * });
   */
  async getLatestNews(count = 10) {
    const articles = await this.getAllNews(100);
    
    // مرتب‌سازی بر اساس تاریخ انتشار / Sort by publish date
    const sorted = articles.sort((a, b) => {
      const dateA = new Date(a.published_at || 0);
      const dateB = new Date(b.published_at || 0);
      return dateB - dateA;
    });
    
    return sorted.slice(0, count);
  }

  /**
   * دریافت آمار اخبار
   * Get news statistics
   * 
   * @returns {Promise<Object>} آمار / Statistics
   * 
   * @example
   * const client = new CryptoNewsClient();
   * const stats = await client.getNewsStatistics();
   * console.log(`Total: ${stats.total}`);
   * console.log(`Positive: ${stats.positive}`);
   */
  async getNewsStatistics() {
    const articles = await this.getAllNews();
    
    const stats = {
      total: articles.length,
      positive: articles.filter(a => a.sentiment === 'positive').length,
      negative: articles.filter(a => a.sentiment === 'negative').length,
      neutral: articles.filter(a => a.sentiment === 'neutral').length,
      sources: new Set(articles.map(a => a.source?.title || '')).size
    };
    
    return stats;
  }
}

// ==============================================================================
// مثال‌های استفاده / Usage Examples
// ==============================================================================

/**
 * مثال ۱: استفاده ساده / Example 1: Basic Usage
 */
async function example1BasicUsage() {
  console.log('='.repeat(60));
  console.log('مثال ۱: دریافت تمام اخبار / Example 1: Get All News');
  console.log('='.repeat(60));
  
  const client = new CryptoNewsClient();
  const articles = await client.getAllNews(10);
  
  console.log(`\nتعداد مقالات / Number of articles: ${articles.length}\n`);
  
  articles.slice(0, 5).forEach((article, i) => {
    console.log(`${i + 1}. ${article.title || 'No title'}`);
    console.log(`   منبع / Source: ${article.source?.title || 'Unknown'}`);
    console.log(`   احساسات / Sentiment: ${article.sentiment || 'neutral'}`);
    console.log('');
  });
}

/**
 * مثال ۲: فیلتر بر اساس احساسات / Example 2: Sentiment Filtering
 */
async function example2SentimentFiltering() {
  console.log('='.repeat(60));
  console.log('مثال ۲: فیلتر اخبار مثبت / Example 2: Positive News Filter');
  console.log('='.repeat(60));
  
  const client = new CryptoNewsClient();
  const positiveNews = await client.getNewsBySentiment('positive', 50);
  
  console.log(`\nاخبار مثبت / Positive news: ${positiveNews.length}\n`);
  
  positiveNews.slice(0, 3).forEach(article => {
    console.log(`✓ ${article.title || 'No title'}`);
    console.log(`  ${(article.content || '').substring(0, 100)}...`);
    console.log('');
  });
}

/**
 * مثال ۳: جستجو با کلمه کلیدی / Example 3: Keyword Search
 */
async function example3KeywordSearch() {
  console.log('='.repeat(60));
  console.log('مثال ۳: جستجوی بیت‌کوین / Example 3: Bitcoin Search');
  console.log('='.repeat(60));
  
  const client = new CryptoNewsClient();
  const bitcoinNews = await client.searchNews('bitcoin');
  
  console.log(`\nمقالات مرتبط با بیت‌کوین / Bitcoin articles: ${bitcoinNews.length}\n`);
  
  bitcoinNews.slice(0, 5).forEach(article => {
    console.log(`• ${article.title || 'No title'}`);
  });
}

/**
 * مثال ۴: آمار اخبار / Example 4: News Statistics
 */
async function example4Statistics() {
  console.log('='.repeat(60));
  console.log('مثال ۴: آمار اخبار / Example 4: Statistics');
  console.log('='.repeat(60));
  
  const client = new CryptoNewsClient();
  const stats = await client.getNewsStatistics();
  
  console.log('\n📊 آمار / Statistics:');
  console.log(`   مجموع مقالات / Total: ${stats.total}`);
  console.log(`   مثبت / Positive: ${stats.positive} (${(stats.positive/stats.total*100).toFixed(1)}%)`);
  console.log(`   منفی / Negative: ${stats.negative} (${(stats.negative/stats.total*100).toFixed(1)}%)`);
  console.log(`   خنثی / Neutral: ${stats.neutral} (${(stats.neutral/stats.total*100).toFixed(1)}%)`);
  console.log(`   منابع / Sources: ${stats.sources}`);
}

/**
 * مثال ۵: آخرین اخبار / Example 5: Latest News
 */
async function example5LatestNews() {
  console.log('='.repeat(60));
  console.log('مثال ۵: آخرین اخبار / Example 5: Latest News');
  console.log('='.repeat(60));
  
  const client = new CryptoNewsClient();
  const latest = await client.getLatestNews(5);
  
  console.log('\n🕒 آخرین اخبار / Latest news:\n');
  
  latest.forEach((article, i) => {
    const published = article.published_at || '';
    const timeStr = published ? new Date(published).toLocaleString() : 'Unknown time';
    
    console.log(`${i + 1}. ${article.title || 'No title'}`);
    console.log(`   زمان / Time: ${timeStr}`);
    console.log('');
  });
}

/**
 * مثال ۶: فیلتر پیشرفته / Example 6: Advanced Filtering
 */
async function example6AdvancedFiltering() {
  console.log('='.repeat(60));
  console.log('مثال ۶: فیلتر ترکیبی / Example 6: Combined Filters');
  console.log('='.repeat(60));
  
  const client = new CryptoNewsClient();
  
  // دریافت اخبار مثبت درباره اتریوم
  // Get positive news about Ethereum
  const allNews = await client.getAllNews(100);
  
  const filtered = allNews.filter(article => {
    const isPositive = article.sentiment === 'positive';
    const isEthereum = (article.title || '').toLowerCase().includes('ethereum');
    return isPositive && isEthereum;
  });
  
  console.log(`\nاخبار مثبت درباره اتریوم / Positive Ethereum news: ${filtered.length}\n`);
  
  filtered.slice(0, 3).forEach(article => {
    console.log(`✓ ${article.title || 'No title'}`);
    console.log(`  منبع / Source: ${article.source?.title || 'Unknown'}`);
    console.log('');
  });
}

/**
 * تابع اصلی / Main function
 */
async function main() {
  console.log('\n' + '='.repeat(60));
  console.log('نمونه‌های استفاده از API اخبار کریپتو');
  console.log('Crypto News API Usage Examples');
  console.log('='.repeat(60) + '\n');
  
  try {
    // اجرای تمام مثال‌ها / Run all examples
    await example1BasicUsage();
    console.log('\n');
    
    await example2SentimentFiltering();
    console.log('\n');
    
    await example3KeywordSearch();
    console.log('\n');
    
    await example4Statistics();
    console.log('\n');
    
    await example5LatestNews();
    console.log('\n');
    
    await example6AdvancedFiltering();
    
  } catch (error) {
    console.error('\nخطا / Error:', error.message);
    console.error('لطفاً مطمئن شوید که سرور در حال اجرا است');
    console.error('Please make sure the server is running');
  }
}

// اجرای برنامه اگر به صورت مستقیم فراخوانی شود
// Run the program if executed directly
if (typeof window === 'undefined') {
  // Node.js environment
  main();
} else {
  // Browser environment - export for use
  window.CryptoNewsClient = CryptoNewsClient;
  console.log('CryptoNewsClient class is now available globally');
  console.log('Usage: const client = new CryptoNewsClient();');
}

// Export for ES6 modules
export { CryptoNewsClient };
export default CryptoNewsClient;




















